### DISCUSSION TITLE
GtkSwitch does not fire events

### DISCUSSION THREAD

**Discussion by KiAchim** [GtkSharp/GtkSharp#430] [Q&A] [answered]
Hi everyone,
I'm playing around with Glade 3.38.2 on Linux Mint (Linux Mint 21.3 Virginia base: Ubuntu 22.04 jammy) and build an application with c# using Mono (MonoDevelop Version 7.8.4 (build 2)). I tried several GtkWidgets o.k., but the 'GtkSwitch' does not seem to create any events like state-set or activate. Can anyone confirm that this in _their_ environment?
![Auswahl_353](https://github.com/GtkSharp/GtkSharp/assets/61900737/c5adbf2a-2163-4991-9f3f-b24d4c907133)

**Comment by lytico** [comment]
are you shure that you are using GtkSharp https://github.com/GtkSharp/GtkSharp?
and not https://github.com/mono/gtk-sharp?

show your csproj-file

**Reply by KiAchim** [reply]
Thanks for replying!
Well, as I said, I'm working with Mono. But I installed the NuGet-Package GtkSharp, the summary of which points me to [the github project](https://github.com/GtkSharp/GtkSharp), see ![Pakete hinzufügen_001](https://github.com/GtkSharp/GtkSharp/assets/61900737/35f9fd22-ccb8-4a90-b7bb-e65ca83a6ec7)

I attach a project file: [GtkApp.csproj.pdf](https://github.com/GtkSharp/GtkSharp/files/14866076/GtkApp.csproj.pdf)
Note that I have worked on it since I posted my question, but this should not have affected the project file.
What I found out meanwhile is, that a 'manual', i.e. from within the code but not from glade, registration on button-pressed and button-released does work; I used it as a workaround.

**Comment by lytico** [comment]
i can confirm that the switch-handler is not fired: 

see https://github.com/lytico/GtkSharp/tree/lytico/issues/430_switch_state

doc is here:
https://docs.gtk.org/gtk4/signal.Switch.state-set.html

i have no idea why it's not fired, i'll look further ...

**Comment by lytico** [comment, marked as answer]
if found something here:
https://www.mail-archive.com/search?l=gtkmm-list@gnome.org&q=subject:%22Gtk%5C%3A%5C%3ASwitch+problem%2C+not+being+able+to+handle+the+signal+state_set%22&o=newest&f=1 :

> this signal is intended for changing the implementation of how/when state changes are applied, not observing the state.

Activate is wrong too:

https://docs.gtk.org/gtk3/signal.Switch.activate.html:
> Applications should never connect to this signal, but use the [GtkSwitch:active](https://docs.gtk.org/gtk3/property.Switch.active.html) property.

so then, only ButtonReleaseEvent is left to observe, as shown in samples:

https://github.com/GtkSharp/GtkSharp/blob/b7303616129ab5a0ca64def45649ab522d83fa4a/Source/Samples/Sections/Widgets/SwitchSection.cs#L18-L22

i am aware that using Button*-events got a lot of critics on stackoverflow, 
eg. here:

https://stackoverflow.com/questions/7120940/gtk-switch-activate-signal-not-firing
https://stackoverflow.com/questions/23700654/gtk-switch-get-active-strange-working

**Reply by KiAchim** [reply]
Thank you for your reply. Basically the 'SwitchSection.cs' snipped given by you shows the workaround I also found and applied.
It works somewhat. I'd like to investigate in the solutions of the [stackoverflow issue 7120940](https://stackoverflow.com/questions/7120940/gtk-switch-activate-signal-not-firing) also given by you. A comment in one of the answers titled 'Actually, a better way is connect to the notify::active event.' was :
«Note to **_glade user:_** I spent a long time figuring how to register a handler for "listen::active", so this might be useful for other glade users. If you're using glade, you'll notice that there does not seem to be **_"notify::active"_** in the signals list, this is because it's a GObject signal "listen", use the "Detail" column to add the second parameter "active" to register a handler for "listen::active".»
This is what I followed:
![Auswahl_354](https://github.com/GtkSharp/GtkSharp/assets/61900737/c3254b4c-29c4-4ef3-9e19-098c3be9e5c6)

**_Can you tell me the signature of the Event-Handler I have to implement to handle notify::active ?_** 

public void on_switch_OnOff_active_notify(object sender, EventArgs args) does not work.
Builder.Autoconnect does not complain, but again, there is no such event triggered.

**Reply by lytico** [reply]
see: https://docs.gtk.org/gtk3/signal.Switch.activate.html :
>  Applications should never connect to this signal

i tried it here, and it doesn't fire:

https://github.com/lytico/GtkSharp/blob/3df4cea1dab6147eedce76fd8bb66e9fb8756c48/Source/Samples/Sections/Widgets/SwitchSection.cs#L34-L35

So the tip in [stackoverflow issue 7120940](https://stackoverflow.com/questions/7120940/gtk-switch-activate-signal-not-firing) seems not to work.

**Reply by KiAchim** [reply]
Hmm. A pity. But o.k. there is the workaround.
One closing question. Is there an actual documentation of the c# API of Gtk? The [one I found](http://docs.go-mono.com/?link=N%3aGtk)  is quite out of date. (its quite hard to translate the C++ signatures to c#)
Thanks for your support!

**Reply by lytico** [reply]
> Is there an actual documentation of the c# API of Gtk?

no. i always use https://docs.gtk.org/gtk3 or the sources. GtkSharp just a wrapper, so the naming is just c#ified.
